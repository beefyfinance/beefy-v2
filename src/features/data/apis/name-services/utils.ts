import { uniq } from 'lodash-es';
import { type Address, getAddress, type Hash, isAddress, numberToHex } from 'viem';
import { namehash, normalize } from 'viem/ens';
import { ZERO_ADDRESS } from '../../../../helpers/addresses.ts';
import type { ChainId } from '../../entities/chain.ts';

export function hashDomain(domain: string): Hash {
  return namehash(domain);
}

export function normalizeDomain(domain: string): string {
  return normalize(domain);
}

export function normalizeAndHashDomain(domain: string): Hash {
  return hashDomain(normalizeDomain(domain));
}

export function normalizeAddress(address: unknown): Address | undefined {
  if (typeof address === 'string' && address.length === 42 && address !== ZERO_ADDRESS) {
    if (isAddress(address, { strict: false })) {
      return getAddress(address);
    }
  }

  return undefined;
}

export function getAllChainsFromTldToChain(tldToChain: Record<string, ChainId[]>): ChainId[] {
  return uniq(Object.values(tldToChain).flat());
}

/**
 * ENSIP-11: EVM compatible Chain Address Resolution
 * @see https://docs.ens.domains/ensip/11/
 **/
export function chainIdToCoinType(chainId: number): number {
  return (0x80000000 | chainId) >>> 0;
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function coinTypeToReverseDomain(coinType: number): string {
  switch (coinType) {
    case 60:
      return 'addr.reverse';
    case 0x8000_0000:
      return 'default.reverse';
    default: {
      if (coinType < 0 || coinType > 0xffff_ffff) {
        throw new Error(`Invalid coin type: ${coinType}`);
      }
      return `${numberToHex(coinType, { size: 4 }).slice(2)}.reverse`;
    }
  }
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function chainIdToReverseDomain(chainId: number): string {
  return coinTypeToReverseDomain(chainIdToCoinType(chainId));
}
