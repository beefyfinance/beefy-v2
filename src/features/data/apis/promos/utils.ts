export function extractChainId(path: string): string {
  const match = path.match(/.+\/chain\/(.+).json$/);
  const chainId = match?.[1];
  if (chainId) {
    return chainId;
  }
  throw new Error(`Could not extract chain id from path: ${path}`);
}
