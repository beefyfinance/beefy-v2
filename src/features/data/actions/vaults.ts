import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import { getBeefyApi, getConfigApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { FeaturedVaultConfig, VaultConfig } from '../apis/config-types';
import type { BeefyVaultZapSupportResponse } from '../apis/beefy/beefy-api';
import { featureFlag_zapSupportOverrides } from '../utils/feature-flags';

// given the list of vaults is pulled from some api at some point
// we use the api to create an action
// this action should return just enough data for the state to work with

export interface FulfilledAllVaultsPayload {
  byChainId: {
    [chainId: ChainEntity['id']]: VaultConfig[];
  };
  state: BeefyState;
}
export const fetchAllVaults = createAsyncThunk<
  FulfilledAllVaultsPayload,
  void,
  { state: BeefyState }
>('vaults/fetchAllVaults', async (_, { getState }) => {
  const api = getConfigApi();
  const vaults = await api.fetchAllVaults();
  return { byChainId: vaults, state: getState() };
});

export interface FulfilledFeaturedVaultsPayload {
  byVaultId: FeaturedVaultConfig;
}
export const fetchFeaturedVaults = createAsyncThunk<FulfilledFeaturedVaultsPayload>(
  'vaults/fetchFeaturedVaults',
  async () => {
    const api = getConfigApi();
    const featuredVaults = await api.fetchFeaturedVaults();
    return { byVaultId: featuredVaults };
  }
);

export type FulfilledFetchVaultsZapSupportPayload = {
  byVaultId: BeefyVaultZapSupportResponse;
};

export const fetchVaultsZapSupport = createAsyncThunk<
  FulfilledFetchVaultsZapSupportPayload,
  void,
  { state: BeefyState }
>('vaults/fetchVaultsZapSupport', async (_, { getState }) => {
  const api = getBeefyApi();
  const vaultsZapSupport = await api.getVaultZapSupport();
  const overrides = featureFlag_zapSupportOverrides();

  for (const kind of ['beefy', 'oneInch'] as const) {
    const value = overrides[kind];
    let ids = [];
    if (value === 'all') {
      ids = getState().entities.vaults.allIds;
    } else if (Array.isArray(value) && value.length > 0) {
      ids = value;
    }
    for (const id of ids) {
      vaultsZapSupport[id] = [...(vaultsZapSupport[id] || []), kind];
    }
  }

  return { byVaultId: vaultsZapSupport };
});

type FulfilledVaultsLastHarvestPayload = {
  byVaultId: { [vaultId: VaultConfig['id']]: number };
};

export const fetchVaultsLastHarvests = createAsyncThunk<FulfilledVaultsLastHarvestPayload>(
  'vaults/last-harvest',
  async () => {
    const api = getBeefyApi();
    const vaults = await api.getVaultLastHarvest();
    return { byVaultId: vaults };
  }
);
