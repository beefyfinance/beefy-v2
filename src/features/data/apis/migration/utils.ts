import { createAsyncThunk } from '@reduxjs/toolkit';
import type {
  CommonMigrationUpdateFulfilledPayload,
  MigratorExecuteProps,
  MigratorUpdateProps,
} from './migration-types';
import type { BeefyState } from '../../../../redux-types';
import { selectVaultById } from '../../selectors/vaults';
import { selectTokenByAddress } from '../../selectors/tokens';
import { selectUserBalanceToMigrateByVaultId } from '../../selectors/migration';
import type { Step } from '../../reducers/wallet/stepper';
import { walletActions } from '../../actions/wallet-actions';
import { startStepperWithSteps } from '../../actions/stepper';
import { isTokenErc20 } from '../../entities/token';
import { selectAllowanceByTokenAddress } from '../../selectors/allowances';
import type { VaultEntity } from '../../entities/vault';
import BigNumber from 'bignumber.js';
import type Web3 from 'web3';
import { selectChainById } from '../../selectors/chains';
import { getWalletConnectionApi, getWeb3Instance } from '../instances';

export function buildFetchBalance(
  id: string,
  fetchBalance: (
    vault: VaultEntity,
    web3: Web3,
    walletAddress: string,
    state: BeefyState
  ) => Promise<string>
) {
  return createAsyncThunk<
    CommonMigrationUpdateFulfilledPayload,
    MigratorUpdateProps,
    { state: BeefyState }
  >(`migration/${id}/update`, async ({ vaultId, walletAddress }, { getState }) => {
    const state = getState();
    const vault = selectVaultById(state, vaultId);
    const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
    const chain = selectChainById(state, vault.chainId);
    const web3 = await getWeb3Instance(chain);

    const balance = await fetchBalance(vault, web3, walletAddress, state);
    const fixedBalance = new BigNumber(balance).shiftedBy(-depositToken.decimals);
    console.debug(id, vault.id, fixedBalance.toNumber());
    return { vaultId, walletAddress, balance: fixedBalance, migrationId: id };
  });
}

export function buildExecute(
  id: string,
  unstakeCall: (
    vault: VaultEntity,
    web3: Web3,
    amount: BigNumber,
    state: BeefyState
    // eslint-disable-next-line
  ) => Promise<any>
) {
  return createAsyncThunk<void, MigratorExecuteProps, { state: BeefyState }>(
    `migration/${id}/execute`,
    async ({ vaultId, t, migrationId }, { getState, dispatch }) => {
      const steps: Step[] = [];
      const state = getState();
      const vault = selectVaultById(state, vaultId);
      const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
      const walletApi = await getWalletConnectionApi();
      const web3 = await walletApi.getConnectedWeb3Instance();
      const { balance } = selectUserBalanceToMigrateByVaultId(state, vaultId, migrationId);

      const call = await unstakeCall(vault, web3, balance, state);

      steps.push({
        step: 'migration',
        message: t('Vault-MigrationStart'),
        action: walletActions.migrateUnstake(
          call,
          vault,
          balance.shiftedBy(depositToken.decimals),
          migrationId
        ),
        pending: false,
        extraInfo: { vaultId },
      });

      if (isTokenErc20(depositToken)) {
        const allowance = selectAllowanceByTokenAddress(
          state,
          depositToken.chainId,
          depositToken.address,
          vault.earnContractAddress
        );
        if (allowance.lt(balance)) {
          steps.push({
            step: 'approve',
            message: t('Vault-ApproveMsg'),
            action: walletActions.approval(depositToken, vault.earnContractAddress),
            pending: false,
          });
        }
      }

      steps.push({
        step: 'deposit',
        message: t('Vault-TxnConfirm', { type: t('Deposit-noun') }),
        action: walletActions.deposit(vault, balance, true),
        pending: false,
        extraInfo: { vaultId: vault.id },
      });

      dispatch(startStepperWithSteps(steps, vault.chainId));
    }
  );
}
