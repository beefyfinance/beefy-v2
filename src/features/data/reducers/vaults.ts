import { createSlice } from '@reduxjs/toolkit';
import type BigNumber from 'bignumber.js';
import type { Draft } from 'immer';
import { isEmpty, mapValues, sortBy } from 'lodash-es';
import { safetyScoreNum } from '../../../helpers/safetyScore';
import type { BeefyState } from '../../../redux-types';
import { fetchAllContractDataByChainAction } from '../actions/contract-data';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens';
import {
  fetchAllVaults,
  fetchFeaturedVaults,
  fetchVaultsZapSupport,
  fetchVaultsLastHarvests,
} from '../actions/vaults';
import type { FetchAllContractDataResult } from '../apis/contract-data/contract-data-types';
import type { ChainEntity } from '../entities/chain';
import type { VaultEntity, VaultGov, VaultStandard } from '../entities/vault';
import type { NormalizedEntity } from '../utils/normalized-entity';
import type { FeaturedVaultConfig, VaultConfig } from '../apis/config-types';

/**
 * State containing Vault infos
 */
export type VaultsState = NormalizedEntity<VaultEntity> & {
  byChainId: {
    [chainId: ChainEntity['id']]: {
      // add quick access arrays
      // doesn't need to be by chain but it's convenient
      // for when we load by chain
      allActiveIds: VaultEntity['id'][];
      allRetiredIds: VaultEntity['id'][];

      // used to find a vault by it's token for balance stuff
      standardVault: {
        byDepositTokenAddress: {
          [address: string]: VaultEntity['id'][];
        };
        byEarnedTokenAddress: {
          [address: string]: VaultEntity['id'];
        };
      };
      govVault: {
        byDepositTokenAddress: {
          [address: string]: VaultEntity['id'][];
        };
      };
    };
  };

  /**
   * pricePerFullShare is how you find out how much your mooTokens
   * (shares) represent in term of the underlying asset
   *
   * So if you deposit 1 BIFI you will get, for example 0.95 mooBIFI,
   * with a ppfs of X, if you multiply your mooBIIFI * ppfs you get your amount in BIFI
   *
   * That value is fetched from the smart contract upon loading
   **/
  contractData: {
    byVaultId: {
      [vaultId: VaultEntity['id']]: {
        strategyAddress: string;
        pricePerFullShare: BigNumber;
      };
    };
  };

  /**
   * We want to know if the vault is featured or not
   */
  featuredVaults: FeaturedVaultConfig;

  /**
   * Which zaps are enabled for each vault
   */
  zapSupportById: {
    [vaultId: VaultEntity['id']]: {
      beefy: boolean;
      oneInch: boolean;
    };
  };

  lastHarvestById: {
    [vaultId: VaultEntity['id']]: number;
  };
};
export const initialVaultsState: VaultsState = {
  byId: {},
  allIds: [],
  byChainId: {},
  contractData: { byVaultId: {} },
  featuredVaults: {},
  zapSupportById: {},
  lastHarvestById: {},
};

export const vaultsSlice = createSlice({
  name: 'vaults',
  initialState: initialVaultsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchFeaturedVaults.fulfilled, (sliceState, action) => {
      sliceState.featuredVaults = action.payload.byVaultId;
    });

    builder.addCase(fetchAllVaults.fulfilled, (sliceState, action) => {
      const initialVaultAmount = sliceState.allIds.length;
      for (const [chainId, vaults] of Object.entries(action.payload.byChainId)) {
        for (const vault of vaults) {
          addVaultToState(action.payload.state, sliceState, chainId, vault);
        }
      }

      // If new vaults were added, apply default sorting
      if (sliceState.allIds.length !== initialVaultAmount) {
        sliceState.allIds = sortBy(sliceState.allIds, id => {
          return -sliceState.byId[id].createdAt;
        });
      }
    });

    builder.addCase(fetchAllContractDataByChainAction.fulfilled, (sliceState, action) => {
      addContractDataToState(sliceState, action.payload.data);
    });

    builder.addCase(fetchVaultsLastHarvests.fulfilled, (sliceState, action) => {
      for (const [vaultId, lastHarvest] of Object.entries(action.payload.byVaultId)) {
        // time in milliseconds since unix epoch
        sliceState.lastHarvestById[vaultId] = lastHarvest * 1000;
      }
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        addContractDataToState(sliceState, action.payload.contractData);
      }
    );

    builder.addCase(fetchVaultsZapSupport.fulfilled, (sliceState, action) => {
      sliceState.zapSupportById = mapValues(action.payload.byVaultId, support => {
        return {
          beefy: support.includes('beefy'),
          oneInch: support.includes('oneInch'),
        };
      });
    });
  },
});

function addContractDataToState(
  sliceState: Draft<VaultsState>,
  contractData: FetchAllContractDataResult
) {
  for (const vaultContractData of contractData.standardVaults) {
    const vaultId = vaultContractData.id;

    // only update it if needed
    if (sliceState.contractData.byVaultId[vaultId] === undefined) {
      sliceState.contractData.byVaultId[vaultId] = {
        pricePerFullShare: vaultContractData.pricePerFullShare,
        strategyAddress: vaultContractData.strategy,
      };
    }

    if (
      !sliceState.contractData.byVaultId[vaultId].pricePerFullShare.isEqualTo(
        vaultContractData.pricePerFullShare
      )
    ) {
      sliceState.contractData.byVaultId[vaultId].pricePerFullShare =
        vaultContractData.pricePerFullShare;
    }
    if (sliceState.contractData.byVaultId[vaultId].strategyAddress !== vaultContractData.strategy) {
      sliceState.contractData.byVaultId[vaultId].strategyAddress = vaultContractData.strategy;
    }
  }
}

function addVaultToState(
  state: BeefyState,
  sliceState: Draft<VaultsState>,
  chainId: ChainEntity['id'],
  apiVault: VaultConfig
) {
  // we already know this vault
  if (apiVault.id in sliceState.byId) {
    return;
  }
  const score = getVaultSafetyScore(state, chainId, apiVault);

  if (apiVault.isGovVault) {
    const vault: VaultGov = {
      id: apiVault.id,
      name: apiVault.name,
      isGovVault: true,
      depositTokenAddress: apiVault.tokenAddress,
      earnedTokenAddress: apiVault.earnedTokenAddress,
      earnContractAddress: apiVault.earnContractAddress,
      excludedId: apiVault.excluded || null,
      chainId: chainId,
      status: apiVault.status as VaultGov['status'],
      platformId: apiVault.platformId,
      safetyScore: score,
      assetIds: apiVault.assets || [],
      type: 'single',
      risks: apiVault.risks || [],
      buyTokenUrl: apiVault.buyTokenUrl || null,
      addLiquidityUrl: null,
      removeLiquidityUrl: null,
      depositFee: apiVault.depositFee ?? '0%',
      createdAt: apiVault.createdAt ?? 0,
      retireReason: apiVault.retireReason,
      pauseReason: apiVault.pauseReason,
    };

    sliceState.byId[vault.id] = vault;
    sliceState.allIds.push(vault.id);
    if (sliceState.byChainId[vault.chainId] === undefined) {
      sliceState.byChainId[vault.chainId] = {
        allActiveIds: [],
        allRetiredIds: [],
        standardVault: {
          byEarnedTokenAddress: {},
          byDepositTokenAddress: {},
        },
        govVault: {
          byDepositTokenAddress: {},
        },
      };
    }

    const vaultState = sliceState.byChainId[vault.chainId];
    if (apiVault.status === 'eol' || apiVault.status === 'paused') {
      vaultState.allRetiredIds.push(vault.id);
    } else {
      vaultState.allActiveIds.push(vault.id);
    }

    if (!vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()]) {
      vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()] = [];
    }
    vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()].push(
      vault.id
    );
  } else {
    const vault: VaultStandard = {
      id: apiVault.id,
      name: apiVault.name,
      isGovVault: false,
      depositTokenAddress: apiVault.tokenAddress ?? 'native',
      earnContractAddress: apiVault.earnContractAddress,
      earnedTokenAddress: apiVault.earnedTokenAddress,
      strategyTypeId: apiVault.strategyTypeId,
      chainId: chainId,
      platformId: apiVault.platformId,
      status: apiVault.status as VaultStandard['status'],
      type: !apiVault.assets ? 'single' : apiVault.assets.length > 1 ? 'lps' : 'single',
      safetyScore: score,
      assetIds: apiVault.assets || [],
      risks: apiVault.risks || [],
      buyTokenUrl: apiVault.buyTokenUrl || null,
      addLiquidityUrl: apiVault.addLiquidityUrl || null,
      removeLiquidityUrl: apiVault.removeLiquidityUrl || null,
      depositFee: apiVault.depositFee ?? '0%',
      createdAt: apiVault.createdAt ?? 0,
      retireReason: apiVault.retireReason,
      pauseReason: apiVault.pauseReason,
      migrationIds: apiVault.migrationIds,
    };
    // redux toolkit uses immer by default so we can
    // directly modify the state as usual
    sliceState.byId[vault.id] = vault;
    sliceState.allIds.push(vault.id);
    if (sliceState.byChainId[vault.chainId] === undefined) {
      sliceState.byChainId[vault.chainId] = {
        allActiveIds: [],
        allRetiredIds: [],
        standardVault: {
          byEarnedTokenAddress: {},
          byDepositTokenAddress: {},
        },
        govVault: {
          byDepositTokenAddress: {},
        },
      };
    }
    const vaultState = sliceState.byChainId[vault.chainId];
    if (apiVault.status === 'eol' || apiVault.status === 'paused') {
      vaultState.allRetiredIds.push(vault.id);
    } else {
      vaultState.allActiveIds.push(vault.id);
    }

    if (!vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()]) {
      vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()] = [];
    }
    vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()].push(
      vault.id
    );
    vaultState.standardVault.byEarnedTokenAddress[vault.earnedTokenAddress.toLowerCase()] =
      vault.id;
  }
}

function getVaultSafetyScore(
  state: BeefyState,
  chainId: ChainEntity['id'],
  apiVault: VaultConfig
): number {
  let score = 0;
  if (!isEmpty(apiVault.risks)) {
    score = safetyScoreNum(apiVault.risks);
  }

  return score;
}
