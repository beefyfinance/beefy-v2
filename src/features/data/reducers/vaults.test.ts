import { fetchAllVaults, FulfilledAllVaultsPayload } from '../actions/vaults';
import {
  fetchAllContractDataByChainAction,
  FetchAllContractDataFulfilledPayload,
} from '../actions/contract-data';
import { vaultsSlice, initialVaultsState } from './vaults';
import { getBeefyTestingStore } from '../utils/test-utils';
import BigNumber from 'bignumber.js';

describe('Vaults slice tests', () => {
  it('should update state on fulfilled vault list', () => {
    const payload: FulfilledAllVaultsPayload = {
      harmony: [
        // have one gov vault
        {
          id: 'one-bifi-gov',
          logo: 'single-assets/BIFI.png',
          name: 'BIFI Earnings Pool',
          token: 'BIFI',
          tokenDescription: 'Beefy.Finance',
          tokenAddress: '0x6ab6d61428fde76768d7b45d8bfeec19c6ef91a8',
          tokenDecimals: 18,
          tokenDescriptionUrl:
            'https://docs.beefy.finance/moo/ecosystem/bifi-token/tokenomics-and-governance',
          earnedToken: 'ONE',
          earnedTokenAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
          earnContractAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
          poolAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
          pricePerFullShare: 1,
          isGovVault: true,
          tvl: 0,
          oracle: 'tokens',
          oracleId: 'BIFI',
          oraclePrice: 0,
          status: 'active',
          platform: 'Beefy.Finance',
          assets: ['BIFI'],
          risks: [
            'COMPLEXITY_LOW',
            'BATTLE_TESTED',
            'IL_NONE',
            'MCAP_SMALL',
            'AUDIT',
            'CONTRACTS_VERIFIED',
          ],
          stratType: 'SingleStake',
          callFee: 0.5,
          withdrawalFee: '0%',
          createdAt: 1623706144,
          network: 'harmony',
        },

        // have one lp vault
        {
          id: 'sushi-one-1ygg-1eth',
          name: '1YGG-1ETH LP',
          token: '1YGG-1ETH LP',
          tokenDescription: 'Sushi',
          tokenAddress: '0x40112850EFd90e9e17b56de35d86BdFf9f4d07Fd',
          tokenDecimals: 18,
          tokenDescriptionUrl: '#',
          earnedToken: 'mooSushiYGG-ETH',
          earnedTokenAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
          earnContractAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
          pricePerFullShare: 1,
          tvl: 0,
          oracle: 'lps',
          oracleId: 'sushi-one-1ygg-1eth',
          oraclePrice: 0,
          status: 'active',
          platform: 'Sushi',
          assets: ['YGG', 'ETH'],
          callFee: 0.25,
          withdrawalFee: '0%',
          addLiquidityUrl:
            'https://app.sushi.com/add/0x6983D1E6DEf3690C4d616b13597A09e6193EA013/0x63cf309500d8be0B9fDB8F1fb66C821236c0438c',
          buyTokenUrl:
            'https://app.sushi.com/swap?inputCurrency=0x6983D1E6DEf3690C4d616b13597A09e6193EA013&outputCurrency=0x63cf309500d8be0B9fDB8F1fb66C821236c0438c',
          stratType: 'StratLP',
          network: 'harmony',
          risks: [
            'COMPLEXITY_LOW',
            'BATTLE_TESTED',
            'IL_LOW',
            'MCAP_MEDIUM',
            'AUDIT',
            'CONTRACTS_VERIFIED',
          ],
        },

        // one retired vault
        {
          id: 'banana-banana-eol',
          logo: 'degens/BANANA.svg',
          name: 'BANANA',
          token: 'BANANA',
          tokenDescription: 'ApeSwap',
          tokenAddress: '0x603c7f932ED1fc6575303D8Fb018fDCBb0f39a95',
          tokenDecimals: 18,
          tokenDescriptionUrl: '#',
          earnedToken: 'mooApeBANANA',
          earnedTokenAddress: '0xD307e7CC6a302046b7D91D83aa4B8324cFB7a786',
          earnContractAddress: '0xD307e7CC6a302046b7D91D83aa4B8324cFB7a786',
          pricePerFullShare: 1,
          tvl: 0,
          oracle: 'tokens',
          oracleId: 'BANANA',
          oraclePrice: 0,
          status: 'eol',
          platform: 'ApeSwap',
          assets: ['BANANA'],
          buyTokenUrl:
            'https://dex.apeswap.finance/#/swap?inputCurrency=0x603c7f932ED1fc6575303D8Fb018fDCBb0f39a95&outputCurrency=0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c',
          stratType: 'SingleStake',
          network: 'bsc',
        },
      ],
    };
    const action = { type: fetchAllVaults.fulfilled, payload: payload };
    const state = vaultsSlice.reducer(initialVaultsState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = Object.values(state.byId)[0];
    const newState = vaultsSlice.reducer(state, action);
    const afterReDispatch = newState.byId[beforeReDispatch.id];
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should update state on fulfilled vault list', async () => {
    const store = await getBeefyTestingStore();
    const state = store.getState();

    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'harmony',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'one-bifi-gov',
            balance: new BigNumber(12),
            pricePerFullShare: new BigNumber(123),
            strategy: 'test',
          },
          {
            id: 'sushi-one-1ygg-1eth',
            balance: new BigNumber(24),
            pricePerFullShare: new BigNumber(456),
            strategy: 'test',
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const sliceState = vaultsSlice.reducer(state.entities.vaults, action);
    // don't snapshot all vaults from the test state
    expect(sliceState.pricePerFullShare).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = sliceState.pricePerFullShare;
    const newState = vaultsSlice.reducer(sliceState, action);
    const afterReDispatch = newState.pricePerFullShare;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });
});
