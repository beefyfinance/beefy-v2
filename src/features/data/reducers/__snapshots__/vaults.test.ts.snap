// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vaults slice tests should handle fuse-fuse config correctly 1`] = `
Object {
  "allIds": Array [
    "fuse-fuse",
  ],
  "byChainId": Object {
    "fuse": Object {
      "allActiveIds": Array [
        "fuse-fuse",
      ],
      "allRetiredIds": Array [],
      "govVault": Object {
        "byOracleId": Object {},
      },
      "standardVault": Object {
        "byEarnTokenId": Object {
          "mooFuse": "fuse-fuse",
        },
        "byOracleId": Object {
          "FUSE": Array [
            "fuse-fuse",
          ],
        },
      },
    },
  },
  "byId": Object {
    "fuse-fuse": Object {
      "addLiquidityUrl": null,
      "assetIds": Array [
        "WFUSE",
      ],
      "buyTokenUrl": "https://app.fuse.fi/#/swap?inputCurrency=FUSE&outputCurrency=0xa722c13135930332Eb3d749B2F0906559D2C5b99",
      "chainId": "fuse",
      "contractAddress": "0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3",
      "createdAt": 1641908745,
      "depositFee": "0%",
      "earnedTokenId": "mooFuse",
      "id": "fuse-fuse",
      "isGovVault": false,
      "logoUri": "single-assets/FUSE.svg",
      "name": "FUSE",
      "oracleId": "FUSE",
      "platformId": "fuse",
      "risks": Array [],
      "safetyScore": 0,
      "status": "active",
      "strategyType": "SingleStake",
      "tags": Array [],
      "type": "single",
      "withdrawalFee": "0%",
    },
  },
  "contractData": Object {
    "byVaultId": Object {},
  },
  "featuredVaults": Object {},
}
`;

exports[`Vaults slice tests should track featured vault list 1`] = `
Object {
  "nope": false,
  "test-vault": true,
}
`;

exports[`Vaults slice tests should update state on fulfilled vault list 1`] = `
Object {
  "allIds": Array [
    "sushi-one-1ygg-1eth",
    "one-bifi-gov",
    "banana-banana-eol",
  ],
  "byChainId": Object {
    "harmony": Object {
      "allActiveIds": Array [
        "one-bifi-gov",
        "sushi-one-1ygg-1eth",
      ],
      "allRetiredIds": Array [
        "banana-banana-eol",
      ],
      "govVault": Object {
        "byOracleId": Object {
          "BIFI": Array [
            "one-bifi-gov",
          ],
        },
      },
      "standardVault": Object {
        "byEarnTokenId": Object {
          "mooApeBANANA": "banana-banana-eol",
          "mooSushiYGG-ETH": "sushi-one-1ygg-1eth",
        },
        "byOracleId": Object {
          "BANANA": Array [
            "banana-banana-eol",
          ],
          "sushi-one-1ygg-1eth": Array [
            "sushi-one-1ygg-1eth",
          ],
        },
      },
    },
  },
  "byId": Object {
    "banana-banana-eol": Object {
      "addLiquidityUrl": null,
      "assetIds": Array [
        "BANANA",
      ],
      "buyTokenUrl": "https://dex.apeswap.finance/#/swap?inputCurrency=0x603c7f932ED1fc6575303D8Fb018fDCBb0f39a95&outputCurrency=0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c",
      "chainId": "harmony",
      "contractAddress": "0xD307e7CC6a302046b7D91D83aa4B8324cFB7a786",
      "createdAt": 1615196201,
      "depositFee": "0%",
      "earnedTokenId": "mooApeBANANA",
      "id": "banana-banana-eol",
      "isGovVault": false,
      "logoUri": "degens/BANANA.svg",
      "name": "BANANA",
      "oracleId": "BANANA",
      "platformId": "apeswap",
      "risks": Array [],
      "safetyScore": 0,
      "status": "eol",
      "strategyType": "SingleStake",
      "tags": Array [
        "eol",
      ],
      "type": "single",
      "withdrawalFee": "0.1%",
    },
    "one-bifi-gov": Object {
      "addLiquidityUrl": null,
      "assetIds": Array [
        "BIFI",
      ],
      "buyTokenUrl": null,
      "chainId": "harmony",
      "createdAt": 1623706144,
      "depositFee": "0%",
      "earnContractAddress": "0x5b96bbaca98d777cb736dd89a519015315e00d02",
      "earnedTokenId": "ONE",
      "excludedId": null,
      "id": "one-bifi-gov",
      "isGovVault": true,
      "logoUri": "single-assets/BIFI.png",
      "name": "BIFI Earnings Pool",
      "oracleId": "BIFI",
      "platformId": "beefy.finance",
      "risks": Array [
        "COMPLEXITY_LOW",
        "BATTLE_TESTED",
        "IL_NONE",
        "MCAP_SMALL",
        "AUDIT",
        "CONTRACTS_VERIFIED",
      ],
      "safetyScore": 9.399999999999999,
      "status": "active",
      "tags": Array [
        "beefy",
        "low",
      ],
      "type": "single",
      "withdrawalFee": "0%",
    },
    "sushi-one-1ygg-1eth": Object {
      "addLiquidityUrl": "https://app.sushi.com/add/0x6983D1E6DEf3690C4d616b13597A09e6193EA013/0x63cf309500d8be0B9fDB8F1fb66C821236c0438c",
      "assetIds": Array [
        "YGG",
        "ETH",
      ],
      "buyTokenUrl": "https://app.sushi.com/swap?inputCurrency=0x6983D1E6DEf3690C4d616b13597A09e6193EA013&outputCurrency=0x63cf309500d8be0B9fDB8F1fb66C821236c0438c",
      "chainId": "harmony",
      "contractAddress": "0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71",
      "createdAt": 1631279238,
      "depositFee": "0%",
      "earnedTokenId": "mooSushiYGG-ETH",
      "id": "sushi-one-1ygg-1eth",
      "isGovVault": false,
      "logoUri": undefined,
      "name": "1YGG-1ETH LP",
      "oracleId": "sushi-one-1ygg-1eth",
      "platformId": "sushi",
      "risks": Array [
        "COMPLEXITY_LOW",
        "BATTLE_TESTED",
        "IL_LOW",
        "MCAP_MEDIUM",
        "AUDIT",
        "CONTRACTS_VERIFIED",
      ],
      "safetyScore": 9.399999999999999,
      "status": "active",
      "strategyType": "StratLP",
      "tags": Array [
        "low",
      ],
      "type": "lps",
      "withdrawalFee": "0%",
    },
  },
  "contractData": Object {
    "byVaultId": Object {},
  },
  "featuredVaults": Object {},
}
`;

exports[`Vaults slice tests should update state on fulfilled vault list 2`] = `
Object {
  "byVaultId": Object {
    "one-bifi-gov": Object {
      "pricePerFullShare": "123",
      "strategyAddress": "test",
    },
    "sushi-one-1ygg-1eth": Object {
      "pricePerFullShare": "456",
      "strategyAddress": "test",
    },
  },
}
`;
